opentelemetry-sdk/src/metrics/internal/aggregate.rs

```diff
@@ -94,11 +94,14 @@ impl<T: Number<T>> AggregateBuilder<T> {
     /// Wraps the passed in measure with an attribute filtering function.
     fn filter(&self, f: impl Measure<T>) -> impl Measure<T> {
         let filter = self.filter.clone();
-        move |n, mut attrs: AttributeSet| {
-            if let Some(filter) = &filter {
-                attrs.retain(filter.as_ref());
-            }
-            f.call(n, attrs)
+        move |n, attrs: &[KeyValue]| {
+            let filtered_attrs: Vec<KeyValue> = if let Some(filter) = &filter {
+                attrs.iter().filter(|kv| filter(kv)).cloned().collect()
+            } else {
+                // TODO: This is unwanted allocation. We should avoid it.
```

Comment:
Just noticed that you added a TODO comment 😀.

I think [this](https://github.com/open-telemetry/opentelemetry-rust/pull/1977/files#r1699088220) should fix it. Up to you if you want to do it in this PR or not.